modules/apps/transfer/keeper/msg_server_test.go

```diff
@@ -238,3 +239,108 @@ func (suite *KeeperTestSuite) TestUpdateParams() {
 		})
 	}
 }
+
+func (suite *KeeperTestSuite) TestUnwindHops() {
+	var msg *types.MsgTransfer
+	var path *ibctesting.Path
+	denom := types.NewDenom(ibctesting.TestCoin.Denom, types.NewTrace(ibctesting.MockPort, "channel-0"), types.NewTrace(ibctesting.MockPort, "channel-1"))
+	coins := sdk.NewCoins(sdk.NewCoin(denom.IBCDenom(), ibctesting.TestCoin.Amount))
+	testCases := []struct {
+		name         string
+		malleate     func()
+		assertResult func(modified *types.MsgTransfer, err error)
+	}{
+		{
+			"success",
+			func() {
+				suite.chainA.GetSimApp().TransferKeeper.SetDenom(suite.chainA.GetContext(), denom)
+			},
+			func(modified *types.MsgTransfer, err error) {
+				suite.Require().NoError(err, "got unexpected error from unwindHops")
+				msg.SourceChannel = denom.Trace[0].PortId
+				msg.SourcePort = denom.Trace[0].ChannelId
+				msg.Forwarding = types.NewForwarding(false, types.Hop{PortId: denom.Trace[1].PortId, ChannelId: denom.Trace[1].ChannelId})
+				suite.Require().Equal(*msg, *modified, "expected msg and modified msg are different")
+			},
+		},
+		{
+			"success - unwind hops are added to existing hops",
+			func() {
+				suite.chainA.GetSimApp().TransferKeeper.SetDenom(suite.chainA.GetContext(), denom)
+				msg.Forwarding = types.NewForwarding(true, types.Hop{PortId: ibctesting.MockPort, ChannelId: "channel-2"})
+			},
+			func(modified *types.MsgTransfer, err error) {
+				suite.Require().NoError(err, "got unexpected error from unwindHops")
+				msg.SourceChannel = denom.Trace[0].PortId
+				msg.SourcePort = denom.Trace[0].ChannelId
+				msg.Forwarding = types.NewForwarding(false,
+					types.Hop{PortId: denom.Trace[1].PortId, ChannelId: denom.Trace[1].ChannelId},
+					types.Hop{PortId: ibctesting.MockPort, ChannelId: "channel-2"},
+				)
+				suite.Require().Equal(*msg, *modified, "expected msg and modified msg are different")
+			},
+		},
+		{
+			"failure: no denom set on keeper",
+			func() {
+				hexHash := denom.IBCDenom()[len(types.DenomPrefix+"/"):]
+
+				hash, err := types.ParseHexHash(hexHash)
+				suite.Require().NoError(err)
+				suite.chainA.DeleteKey(hash)
+			},
+			func(modified *types.MsgTransfer, err error) {
+				suite.Require().Error(err)
+			},
```

Comment:
I don't think this test is testing what we want it to!

A few things,

- in order to get the `hexHash` we can juse use `denom.Hash()` which returns what you are parsing.
- We don't actually set the denom before this test runs, so we don't need to delete the denom, since it not there to begin with. e.g., if you add `suite.chainA.GetSimApp().TransferKeeper.SetDenom(suite.chainA.GetContext(), denom)`  to the beginning of the malleate, this test will fail.
- The key we want to delete is actually in a prefixed store, not the `ibc` store. 

Sorry I led you down the `DeleteKey` route but actually we need to use a prefixed store for the transfer module and delete it there. This should do the trick

```golang
			func() {
				suite.chainA.GetSimApp().TransferKeeper.SetDenom(suite.chainA.GetContext(), denom)

				storeKey := suite.chainA.GetSimApp().GetKey(types.StoreKey) // types.StoreKey == "transfer"
				kvStore := suite.chainA.GetContext().KVStore(storeKey)

				store := prefix.NewStore(kvStore, types.DenomKey)
				store.Delete(denom.Hash())
			},
```

Reply:
I think it's actually my fault for moving some tests stuff around after our call :)

But I think since we don't set the denom for each test now, we can simply avoid all of this. Please take a look at the new implementation and LMK if it looks ok. I als ochanged the suite.Require to test for the exact error type